src/Sentry/MetricAggregator.cs

```diff
@@ -0,0 +1,273 @@
+        {
+            MetricType.Counter => (string _) => new CounterMetric(key, value, unit.Value, tags, timestamp),
+            MetricType.Gauge => (string _) => new GaugeMetric(key, value, unit.Value, tags, timestamp),
+            MetricType.Distribution => (string _) => new DistributionMetric(key, value, unit.Value, tags, timestamp),
+            MetricType.Set => (string _) => new SetMetric(key, (int)value, unit.Value, tags, timestamp),
+            _ => throw new ArgumentOutOfRangeException(nameof(type), type, "Unknown MetricType")
+        };
+
+        timeBucket.AddOrUpdate(
+            GetMetricBucketKey(type, key, unit.Value, tags),
+            addValuesFactory,
+            (_, metric) =>
+            {
+                metric.Add(value);
+                return metric;
+            }
+        );
+    }
+
+    private async Task RunLoopAsync()
+    {
+        _options.LogDebug("MetricsAggregator Started.");
+
+        using var shutdownTimeout = new CancellationTokenSource();
+        var shutdownRequested = false;
+
+        try
+        {
+            while (!shutdownTimeout.IsCancellationRequested)
+            {
+                // If the cancellation was signaled, run until the end of the queue or shutdownTimeout
+                if (!shutdownRequested)
+                {
+                    try
+                    {
+                        var delay = Task.Delay(_flushInterval, shutdownTimeout.Token).ConfigureAwait(false);
+                        await Task.Delay(_flushInterval, shutdownTimeout.Token).ConfigureAwait(false);
+                    }
+                    // Cancellation requested and no timeout allowed, so exit even if there are more items
+                    catch (OperationCanceledException) when (_options.ShutdownTimeout == TimeSpan.Zero)
+                    {
+                        _options.LogDebug("Exiting immediately due to 0 shutdown timeout.");
+
+                        shutdownTimeout.Cancel();
+
+                        return;
+                    }
+                    // Cancellation requested, scheduled shutdown
+                    catch (OperationCanceledException)
+                    {
+                        _options.LogDebug(
+                            "Shutdown scheduled. Stopping by: {0}.",
+                            _options.ShutdownTimeout);
+
+                        shutdownTimeout.CancelAfterSafe(_options.ShutdownTimeout);
+
+                        shutdownRequested = true;
+                    }
+                }
+
+                // Work with the envelope while it's in the queue
+                foreach (var key in GetFlushableBuckets())
+                {
+                    // TODO: Check if a shutdown request has been made
+                    try
+                    {
+                        _options.LogDebug("Flushing metrics for bucket {0}", key);
+                        var bucket = Buckets[key];
```

Comment:
You have a race condition here that can cause data loss. Happens between this operation and line 201 where you remove it, other threads can write new metrics on this bucket.

You should replace this indexer access with `TryRemove` you used below, and use the reference you extracted out (now discarded with `_`) of the map to do the work here. This also means the `finally` block is redundant

Reply:
I thought about that. Given how the cutoff works, nothing will have been written to the bucket for at least 10-20 seconds by the time the bucket gets flushed... so no race condition.

I figured this way if there was some issue with the `_captureMetrics` call, the bucket wouldn't have been removed so we could maybe try again. Thinking on that now though, we don't do anything to prevent a potential infinite loop in this case... so maybe better off removing it from the Buckets collection and just making a best effort to deliver the metrics to Sentry (we have retry logic etc. in the Transport/Envelope handling logic anyway).